include/mapnik/geometry_is_valid.hpp

```diff
@@ -106,6 +110,140 @@ inline bool is_valid(T const& geom)
     return detail::geometry_is_valid() (geom);
 }
 
+typedef boost::geometry::model::d2::point_xy<double> Point;
+typedef boost::geometry::model::segment<Point> Segment;
+
+template <typename T>
+std::vector<Segment> make_segment_vector(linear_ring<T> ring)
+{
+    std::vector<Segment> result;
+
+    if (ring.size() > 1) {
+        for (int i = 0; i < ring.size()-1; ++i)
+        {
+            Point a( ring.at(i).x, ring.at(i).y );
+            Point b( ring.at(i+1).x, ring.at(i+1).y );
+            Segment s(a, b);
+            result.push_back(std::move(s));
+        }
+    }
+
+    return result;
+
+}
+
+template <typename T>
+inline bool is_valid_rings(polygon<T> const& poly)
+{
+    // If there are no interior rings, then the polygon must be valid
+    if (poly.interior_rings.size() == 0) return true;
+
+    // If there is no exterior rings, and any interior rings have points
+    // then this is not a valid polygon
+    if (poly.exterior_ring.size() == 0) 
+    {
+        for (linear_ring<T> interior_ring : poly.interior_rings) 
+        {
+            if (interior_ring.size() > 0) 
+            {
+                return false;
+            }
+        }
+    }
+
+    // Validate that all the rings are actually rings (i.e. they don't intersect with themselves)
+    // i.e. they're simple polygons.
+    if (boost::geometry::intersects(poly.exterior_ring))
+    {
+        return false;
+    }
+    for (auto ring : poly.interior_rings) 
+    {
+        if (boost::geometry::intersects(ring)) 
+        {
+            return false;
+        }
+    }
+
+    // Now, verify the winding directions of everything:  CCW for the exterior ring, and
+    // CW for any interior rings.
+    if (boost::geometry::area(poly.exterior_ring) < 0) 
+    {
+        return false;
+    }
+    for (auto ring : poly.interior_rings) 
+    {
+        if (boost::geometry::area(ring) > 0) 
+        {
+            return false;
+        }
+    }
+
+    // Quick test to see if the interior rings have at least one point
+    // inside the exterior ring.  This test, combined with the line crossing
+    // test next will ensure that all rings are inside the exterior.
+    // Also, check that rings start/end on the same point
+    for (linear_ring<T> ring : poly.interior_rings) 
+    {
+        if (ring.size() > 0) 
+        {
+            if (!boost::geometry::within(ring.front(), poly.exterior_ring)) 
+            {
+                return false;
+            }
+            if (ring.size() > 1) 
+            {
+                if (ring.front().x != ring.back().x || ring.front().y != ring.back().y) 
+                {
+                    return false;
+                }
+            }
+        }
+    }
+
+    // Then, make sure there are no line intersections between the exterior ring
+    // and any of the interior rings
+    // TODO: implement boost::geometry::crosses, or convert this to Shamos–Hoey
+
+    auto exterior_segments = make_segment_vector(poly.exterior_ring);
```

Comment:
Since linear_ring can be downcast to line_string you should be able to do it and implement crosses. 
https://github.com/mapnik/mapnik/blob/master/include/mapnik/geometry.hpp#L77